runtime/src/in_mem_accounts_index.rs

```diff
@@ -18,15 +18,15 @@ type K = Pubkey;
 #[derive(Debug)]
 pub struct InMemAccountsIndex<T: IsCached> {
     // backing store
-    map: HashMap<Pubkey, AccountMapEntry<T>>,
+    map: RwLock<HashMap<Pubkey, AccountMapEntry<T>>>,
```

Comment:
Yes, you are correct, to be shared, it needs an arc. The HashMap will soon be owned by the CachedBucket struct in a shortly arriving PR. InMemAccountsIndex will have an Arc<CachedBucket>. So, we'll be using that Arc to provide the share ability to the RwLock<Hashmap>. Good observation.

Reply:
The locks are confusing. Entry requires a write lock. get and other methods require a read lock. This is the way the accounts index and its locking from the client api level work already. This was intended to be a simple step that is easily verifiably correct that paves the way for the real locking behavior which is more complicated and coming soon.

Reply:
I think I'm still confused since `storage` is wrapped in an `Arc`. I would think that either both `map` and `storage` are in their own `Arc`, or neither are. I'm assuming I'm missing something though, can you help me understand?

Reply:
Or I guess this is at a stage where there is no sharing yet, so it doesn't matter?

Reply:
Storage will evolve and may disappear from InMemAccountsIndex altogether. Probably InMemAccountsIndex will JUST be a forwarder to methods on the arc to its CachedBucket. InMemAccountsIndex is itself held in Vec<RwLock<InMemAccountsIndex>> by the AccountsIndex. That is the client api's locking api. InMemAccountsIndex allows me to implement the account index's client api (used by AccountsDb, AccountsIndex, etc.). But, the real implementation of each method (keys, get, etc.) will require coordination between the in-mem, hot-cache of index items and the disk buckets. Working between those 2 objects will require locking at that level. Introducing the INTERNAL RwLock<HashMap> in this PR is the beginning to controlling access to the in-mem hot-cache of index items. For now, this HashMap still contains ALL the items and is never flushed or sync'd with disk. Soon, it will be. Right now, storage is providing the shared stats as its only useful data, I think. Soon, there will likely be no stat reporting at all in InMemAccountsIndex and we will just have method forwarders:
```
fn foo(&self) -> bar {
   self.bucket.foo()
}
```
You have to back up a little and look at ownership. And perhaps something isn't exactly right - that will become clear. I'm working a few steps ahead to try to make sure I have high confidence that the current step is reasonably ideal but more importantly a clearly not incorrect step. WRT ownership, an InMemAccountsIndex instance has a single instance (no arc) and is owned by a RwLock owned by the single instance AccountsIndex. An InMemAccountsIndex will have an Arc<CachedBucket>. CachedBucket will have a RwLock<HashMap> and have access to the DiskBuckets (not in master yet). DiskBuckets have their OWN locking. BG threads will be working with Arc<CachedBucket>, RwLock<HashMap> and DiskBuckets (which internally will have Vec<RwLock<DiskBucket>>, corresponding to the bins we already have (but they don't have to correspond - but I intend for them to to keep things simple)).
These are great questions and it is complicated. It should become more clear once we have these next few pieces in place. Some things are temporarily in the wrong place because of making simple, verifiable steps forward.

Reply:
Thank you! Lots of pieces that I haven't yet been able to fully layout in my head yet, but I bet as I re-read this a few more times over the coming days it'll make more sense.